tsl/src/fdw/estimate.c

```diff
@@ -285,53 +136,312 @@ adjust_foreign_grouping_path_cost(PlannerInfo *root, List *pathkeys, double retr
-		 * base foreign relations.
-		 */
-		.rows = rel->rows,
-		.width = rel->reltarget->width,
-	};
+	TsFdwRelInfo *fpinfo = fdw_relinfo_get(foreignrel);
+	double rows;
+	double retrieved_rows;
+	int width;
+	Cost startup_cost;
+	Cost total_cost;
+	Cost run_cost = 0;
 
-	if (IS_JOIN_REL(rel))
+	/* Make sure the core code has set up the relation's reltarget */
+	Assert(foreignrel->reltarget);
+
+	if (IS_JOIN_REL(foreignrel))
 		ereport(ERROR,
 				(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
 				 errmsg("foreign joins are not supported")));
 
 	/*
-	 * We will come here again and again with different set of pathkeys
-	 * that caller wants to cost. We don't need to calculate the cost of
-	 * bare scan each time. Instead, use the costs if we have cached them
-	 * already.
+	 * We don't support join conditions in this mode (hence, no
+	 * parameterized paths can be made).
+	 */
+	Assert(param_join_conds == NIL);
+
+	/*
+	 * In Timescale, we should be relying on local ANALYZE statistics because
+	 * they get updated with remote stats appropriately
+	 *
+	 * We will come here again and again with different set of pathkeys or
+	 * additional post-scan/join-processing steps that caller wants to
+	 * cost.  We don't need to calculate the cost/size estimates for the
+	 * underlying scan, join, or grouping each time.  Instead, use those
+	 * estimates if we have cached them already.
 	 */
 	if (REL_HAS_CACHED_COSTS(fpinfo))
 	{
-		ce.rows = fpinfo->rows;
-		ce.width = fpinfo->width;
-		ce.startup_cost = fpinfo->rel_startup_cost;
-		ce.run_cost = fpinfo->rel_total_cost - fpinfo->rel_startup_cost;
-		ce.retrieved_rows = fpinfo->rel_retrieved_rows;
+		Assert(fpinfo->rel_retrieved_rows >= 0);
+
+		rows = fpinfo->rows;
+		retrieved_rows = fpinfo->rel_retrieved_rows;
+		width = fpinfo->width;
+		startup_cost = fpinfo->rel_startup_cost;
+		run_cost = fpinfo->rel_total_cost - fpinfo->rel_startup_cost;
+
+		/*
+		 * If we estimate the costs of a foreign scan or a foreign join
+		 * with additional post-scan/join-processing steps, the scan or
+		 * join costs obtained from the cache wouldn't yet contain the
+		 * eval costs for the final scan/join target, which would've been
+		 * updated by apply_scanjoin_target_to_paths(); add the eval costs
+		 * now.
+		 */
+		if (fpextra && !IS_UPPER_REL(foreignrel))
+		{
+			/* Shouldn't get here unless we have LIMIT */
+			Assert(fpextra->has_limit);
+			Assert(foreignrel->reloptkind == RELOPT_BASEREL ||
+				   foreignrel->reloptkind == RELOPT_JOINREL);
+			startup_cost += foreignrel->reltarget->cost.startup;
+			run_cost += foreignrel->reltarget->cost.per_tuple * rows;
+		}
+	}
+	else if (IS_JOIN_REL(foreignrel))
+	{
+		TsFdwRelInfo *fpinfo_i;
+		TsFdwRelInfo *fpinfo_o;
+		QualCost join_cost;
+		QualCost remote_conds_cost;
+		double nrows;
+
+		/* Use rows/width estimates made by the core code. */
+		rows = foreignrel->rows;
+		width = foreignrel->reltarget->width;
+
```

Comment:
Lots of stuff not covered, is this because we don't support join pushdown? May be better to remove it then.

Reply:
Why not leave it around for the future when we do support it? @akuzm

Reply:
We could, but I'm afraid that after being dead code for many months, it might end up broken in some subtle ways. And it will also confuse the readers who might think we somehow support the joins. On the other hand, deleting it now and copying it back will also be a pain, so no strong opinion here.

Reply:
Ok, can remove the JOIN related code path.